### PR TITLE
Add Import Functionality to Prov-Infra Projects

### PR DESCRIPTION
**Pull request by BarisSari**
- [x] add a way to upload this data from an xlsx file in django Admin
- [x] Uploads where a row matches an existing "Project ID" (IRM_project_id) should update that row, resulting in the updated_at field updating automatically 
- [x] When no existing Project ID matches, a new object should be created and saved. 
- [x] Any missing column names should result in an error stating which column name is missing.
- [x] The agent/contractor/parties columns should be imported as per https://trello.com/c/0y5Btf2O/285-pre-process-irm-data 
  - [x] Test that `Program Implementing Agent: `, Main Contractor:, and `Principal Agent:` 
    - [x] get picked up from any of the contractor columns
    - [x] get cleaned
    - [x] get assigned to correct column
  - [ ] Test that contractor values with other keys
    - [x] get picked up from any of the contractor columns
    - [x] keep their key (e.g. `Service provider: `)
    - [x] Get placed in the `Other parties` column
    - [x] multiple values are placed on new lines in the Other parties column
- [x] add automated tests run in Travis CI that verify the import functionality (prefer unit tests for small pieces of logic)
- [x] include high level end-to-end tests that actually upload a file via the admin interface, verifying that the change is made at a high level but don't verify all fields using the end-to-end test. See the existing upload test as an example.

### CODE REVIEW

File: budgetportal/provincial_infrastructure_projects.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    "Main Budget Appropriation (Construction Costs)",
+    "Adjustment Budget Appropriation (Construction Costs)",
+    "Main Budget Appropriation (TOTAL)",
+    "Adjustment Budget Appropriation (TOTAL)",
+    "Main Budget Appropriation (Professional Fees)",
+    "Adjustment Budget Appropriation (Professional Fees)",
+    "Main Budget Appropriation (Construction Costs)",
+    "Adjustment Budget Appropriation (Construction Costs)",
+    "Main Budget Appropriation (TOTAL)",
+    "Adjustment Budget Appropriation (TOTAL)",
```
Comment (jbothma) [inline budgetportal/provincial_infrastructure_projects.py]: looks like duplication from row 27-32 which we've had in earlier data

Comment (BarisSari) [reply budgetportal/provincial_infrastructure_projects.py]: yes, that columns were duplicate in the example spreadsheet.

File: budgetportal/tests/test_provincial_infrastructure_projects.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        selenium.find_element_by_link_text('IMPORT').click()
+
+        file_import = selenium.find_element_by_id('id_import_file')
+        financial_year_select = Select(selenium.find_element_by_id('id_financial_year'))
+        file_import.send_keys(os.path.abspath(filename))
+        financial_year_select.select_by_value(str(self.financial_year.id))
+        selenium.find_element_by_css_selector(
+            'input[type="submit"]').click()
+
+        timeout = 2
```
Comment (jbothma) [inline budgetportal/tests/test_provincial_infrastructure_projects.py]: isn't a 2s wait a bit brittle? I'd suggest more like 10s to be sure the page loads properly and is ready for business. Or am I missing something?

To reduce duplication here, you can set the timeout at the top and use implicit waits https://selenium-python.readthedocs.io/waits.html#implicit-waits

File: budgetportal/tests/test_provincial_infrastructure_projects.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        # check values of the first project
+        first_project = ProvInfraProject.objects.get(IRM_project_id=30682)
+        self.assertEqual(first_project.financial_year.slug, "2019-20")
+        self.assertEqual(first_project.project_number, "W/50042423/WS")
+        self.assertEqual(first_project.name, "BLUE JUNIOR SECONDARY SCHOOL")
+        self.assertEqual(first_project.project_expenditure_total, 1315479)
+
+        # check whether null values are imported as null, and 0 imported as 0
+        self.assertEqual(first_project.contracted_construction_end_date, None)
```
Comment (jbothma) [inline budgetportal/tests/test_provincial_infrastructure_projects.py]: couldyou aso verify that a bank numeric column is imported as None? e.g. `Project Expenditure from Previous Financial Years (Construction Costs)`

Comment (BarisSari) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: solved
https://github.com/vulekamali/datamanager/pull/189/commits/705cd02dd74242eca8f4cb2b66d85fab76ce00a5

File: budgetportal/tests/test_provincial_infrastructure_projects.py (hunk 2 above)
Comment (BarisSari) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: solved

https://github.com/vulekamali/datamanager/pull/189/commits/705cd02dd74242eca8f4cb2b66d85fab76ce00a5

File: budgetportal/provincial_infrastructure_projects.py (hunk 1 above)
Comment (BarisSari) [reply budgetportal/provincial_infrastructure_projects.py]: solved

https://github.com/vulekamali/datamanager/pull/189/commits/bdbc08c77a50811eac41d1b4d18493fc71a1c3af

File: budgetportal/tests/test_provincial_infrastructure_projects.py [hunk 4]
```diff
@@ -86,3 +87,88 @@ def test_upload_xlsx_for_prov_infra_projects(self):
...
+
+        self.report_sheet.process()
+        program_implementing_agent = self.report_sheet.output_data_set[
+            "Program Implementing Agent"
+        ]
+        main_contractor = self.report_sheet.output_data_set["Main Contractor"]
+        principal_agent = self.report_sheet.output_data_set["Principal Agent"]
+        other_parties = self.report_sheet.output_data_set["Other parties"]
+
+        self.assertEqual(program_implementing_agent, ["DOPW\nTEST"])
```
Comment (jbothma) [inline budgetportal/tests/test_provincial_infrastructure_projects.py]: how does this happen when the input data has DOPW as `Service Supplier`?

Comment (jbothma) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: oh sorry, there are two rows

File: budgetportal/tests/test_provincial_infrastructure_projects.py [hunk 5]
```diff
@@ -86,3 +87,88 @@ def test_upload_xlsx_for_prov_infra_projects(self):
...
+        self.assertEqual(main_contractor, ["AAAA"])
+        self.assertEqual(principal_agent, ["BBBB"])
+        self.assertEqual(other_parties, [None])
+
+    def test_other_parties(self):
+        """Checks when different keys given in project contractor"""
+
+        # it basically does the same what process()->process_row() does
+        for index, row in enumerate(self.sheet_with_other_keys.data_set):
```
Comment (jbothma) [inline budgetportal/tests/test_provincial_infrastructure_projects.py]: why is this happening? it looks hacky - isn't there a more normal way of making the processing happen?

File: budgetportal/tests/test_provincial_infrastructure_projects.py [hunk 6]
```diff
@@ -86,3 +87,88 @@ def test_upload_xlsx_for_prov_infra_projects(self):
...
+            + [
+                "Service Provider: DOPW",
+                "Program Implementing Agent: TEST",
+                "Service Provider: AAAA",
+                "Principal Agent: BBBB",
+            ]
+            + [None] * 46
+        )
+        dataset_with_other_keys.append(row_with_other_keys)
+        self.sheet_with_other_keys = IRMReportSheet(dataset_with_other_keys)
```
Comment (jbothma) [inline budgetportal/tests/test_provincial_infrastructure_projects.py]: please split `sheet_with_other_keys` and `report_sheet` into two test classes - there's no value in them being the same class so separation will indicate that they are not connected.

File: budgetportal/tests/test_provincial_infrastructure_projects.py (hunk 5 above)
Comment (jbothma) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: That will also improve data locality - where the test data is defined and checked will be close together, and not interleaved with other tests

File: budgetportal/tests/test_provincial_infrastructure_projects.py (hunk 6 above)
Comment (jbothma) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: That will also improve data locality - where the test data is defined and checked will be close together, and not interleaved with other tests

Comment (BarisSari) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: classes split:
https://github.com/vulekamali/datamanager/pull/189/commits/e8a09cf18378c191e260f74fd6ab5249b3ce4cbe

File: budgetportal/tests/test_provincial_infrastructure_projects.py (hunk 5 above)
Comment (BarisSari) [reply budgetportal/tests/test_provincial_infrastructure_projects.py]: changed the loop with process() method
https://github.com/vulekamali/datamanager/pull/189/commits/2a6005d6eec5ea47e4d404aeeb05b287fbd797ce

**Pull request merged**